src/chart/utils/linkedList.utils.ts

```diff
@@ -0,0 +1,135 @@
+export class ListNode<T> {
+	public data: T;
+	public next: ListNode<T> | null;
+
+	constructor(data: T) {
+		this.data = data;
+		this.next = null;
+	}
+}
+/**
+ * Implementation of Linked list data structure.
+ * @param head
+ * @param epsilon
+ */
+export class LinkedList<T> {
+	private head: ListNode<T> | null = null;
+	private length = 0;
+
+	constructor(head?: ListNode<T>) {
+		this.head = head ?? null;
+	}
+
+	public insertAtEnd(data: T): ListNode<T> {
+		const node = new ListNode(data);
+		let current: ListNode<T>;
+
+		if (this.head === null) {
+			this.head = node;
+		} else {
+			current = this.head;
+			while (current.next) {
```

Comment:
you won't need this loop if you have `tail` stored

Reply:
I think the loop is still needed, because tail only stores the last node data and to keep linked list structure we still need to update head.next (and so on) nodes. But tail is still could be useful to easily get access to the last node so I decided to duplicate the received data into tail (this._tail = node)

Reply:
Couldn't make it work, some weird bug happens with the drawings (incorrect removal and addition of nodes), it also happens with remove(node) method so I reversed it back to removeAt(position) method. 😵‍💫